File: taskcenter/src/main/java/com/iwdael/taskcenter/core/Node.java
package com.iwdael.taskcenter.core;

import com.iwdael.taskcenter.annotations.TaskFieldCreator;

import org.jetbrains.annotations.NotNull;

/**
 * @author : iwdael
 * @mail : iwdael@outlook.com
 * @project : https://github.com/iwdael/TaskCenter
 */

public abstract class Node<S, D, C> {
    protected int index = 0;
    protected Node<Object, Object, Object> next;
    protected Node<Object, Object, Object> pre;
    protected final Class<?> type;
    @TaskFieldCreator
    protected final C creator;

    public Node(Class<?> type, C creator) {
        this.type = type;
        this.creator = creator;
    }

    @NotNull
    protected abstract D make(@NotNull S src);

    @NotNull
    protected Node<Object, Object, Object> asNode() {
        return (Node<Object, Object, Object>) this;
    }
}

File: taskcenter/src/main/java/com/iwdael/taskcenter/core/Node.kt
package com.iwdael.taskcenter.core

abstract class Node<I : Any, R : Any> {
    internal var index = 0
    internal var next: Node<Any, Any>? = null
    internal var pre: Node<Any, Any>? = null
    internal fun hasNext() = next != null
    internal fun next() = next!!
    internal abstract fun make(src: I): R
    internal fun asNode() = this as Node<Any, Any>
}